source/adapters/level_zero/v2/memory.cpp

```diff
@@ -51,55 +51,80 @@ void *ur_host_mem_handle_t::getPtr(ur_device_handle_t hDevice) {
   return ptr;
 }
 
+ur_result_t ur_device_mem_handle_t::migrateBufferTo(ur_device_handle_t hDevice,
+                                                    void *src, size_t size) {
+  auto Id = hDevice->Id.value();
+
+  if (!deviceAllocations[Id]) {
+    UR_CALL(hContext->getDefaultUSMPool()->allocate(hContext, hDevice, nullptr,
+                                                    UR_USM_TYPE_DEVICE, size,
+                                                    &deviceAllocations[Id]));
+  }
+
+  auto commandList = hContext->commandListCache.getImmediateCommandList(
+      hDevice->ZeDevice, true,
+      hDevice
+          ->QueueGroup[ur_device_handle_t_::queue_group_info_t::type::Compute]
+          .ZeOrdinal,
+      ZE_COMMAND_QUEUE_MODE_SYNCHRONOUS, ZE_COMMAND_QUEUE_PRIORITY_NORMAL,
+      std::nullopt);
+
+  ZE2UR_CALL(zeCommandListAppendMemoryCopy,
+             (commandList.get(), deviceAllocations[Id], src, size, nullptr, 0,
+              nullptr));
+
+  activeAllocationDevice = hDevice;
+
+  return UR_RESULT_SUCCESS;
+}
+
 ur_device_mem_handle_t::ur_device_mem_handle_t(ur_context_handle_t hContext,
                                                void *hostPtr, size_t size)
     : ur_mem_handle_t_(hContext, size),
-      deviceAllocations(hContext->getPlatform()->getNumDevices()) {
-  // Legacy adapter allocated the memory directly on a device (first on the
-  // contxt) and if the buffer is used on another device, memory is migrated
-  // (depending on an env var setting).
-  //
-  // TODO: port this behavior or figure out if it makes sense to keep the memory
-  // in a host buffer (e.g. for smaller sizes).
+      deviceAllocations(hContext->getPlatform()->getNumDevices()),
+      activeAllocationDevice(nullptr) {
   if (hostPtr) {
-    buffer.assign(reinterpret_cast<char *>(hostPtr),
-                  reinterpret_cast<char *>(hostPtr) + size);
+    auto initialDevice = hContext->getDevices()[0];
```

Comment:
doing the allocation to first device isn't ideal... Are we sure this hostPtr must not be valid for the lifetime of the buffer?
Anyway, let's keep it this way, but I think we might want to create a todo/issue to revisit this decision at some point.

Reply:
Yes, I'm tracking this a part of one of the issues. For now, this behavior (allocating on the first device) is the same as in legacy behavior (without the migration).